### ISSUE TITLE
TP4 decode logits are not reproducible run to run

### ISSUE THREAD

**Issue by lvyufeng** [lvyufeng/PocketLLM#250]
## Summary

On TP4, greedy decode on Qwen3.8-27B does not produce the same tokens on two runs of the **same binary with the same prompt**. Four runs of two identical binaries gave three distinct step-0 tokens, with the top logit moving between 10.42 and 10.94 — far outside fp16 rounding for a 4966-token prompt.

This matters because generated-token comparison is the obvious way to check an accuracy-neutral optimisation end to end, and on this stack it currently is not usable as evidence in either direction.

## Evidence

4966-token prompt, 9 new tokens, `scripts/run_qwen_ascend_tp4.sh`, 64 layers, TP4, `QWEN_PHASE_PROFILE=1`, one process per rank, no other activity on the part. The whole-head pair is one binary; the sliced pair is another. Nothing differs within a pair.

| run | step-0 token | top logit | checksum |
|---|---|---|---|
| whole head, rep 1 | 125017 | 10.6346 | 10.2468 |
| whole head, rep 2 | 142711 | 10.4235 | 10.2133 |
| value-axis slice, rep 1 | 152058 | 10.939 | 10.0803 |
| value-axis slice, rep 2 | 125017 | 10.6803 | 10.2094 |

Three distinct continuations in four runs, and **the two baseline runs disagree with each other**, so the variation is not arm-to-arm. Note also that rep 1 and rep 2 of the *same* kernel produce different tokens while step-1 onwards diverge completely (`125017, 12, 24273, ...` vs `142711, 95726, 6, ...`), which is the expected amplification of a single flipped argmax.

Log line format:

```
generate_step=0 token=125017 token_text=顾名思义 top_token=125017 top_logit=10.6346 checksum=10.2468
```

Only rank 0 prints these, so there is no cross-rank comparison in the output to say whether the ranks agree with each other within a run.

## Suspected area

The runs above had the TP all-reduce **enabled** — the decode profile shows a real `tp_all_reduce` scope at 1033 calls, so the collective path was live. Two candidates, neither confirmed:

- **The comm-stream collective path.** `9238f11` enabled it by default and required an explicit `aclrtSynchronizeStream` at `end_nccl_collective` because CANN 9.0 on first-generation 910 can return from a default-stream event wait before an HCCL in-place output is visible to later aclnn work. That fix addresses visibility, not ordering, so a reduction whose completion is ordered differently run to run would show up exactly like this.
- **Reduction order inside the layers.** Any accumulate-into-shared-buffer path that depends on cross-rank arrival order would also produce this.

Worth ruling out the same way the group-B failures were: run one binary twice with the collective path made deliberately synchronous, and see whether the logits become bit-stable. If they do, the nondeterminism is in the overlap and not in the kernels — and every decode number taken through the overlap path should be labelled accordingly.

## Impact

- Any A/B that uses generated tokens as the correctness criterion on this backend needs run-to-run stability established first, or it will report a spurious regression.
- Both PR #247's arms and the `QWEN_ASCEND_GQA_CUBE=0` control in #249 are affected by the same caveat where they use token output.

## Details

First-generation Ascend 910 (`Short_SoC_version=Ascend910`), CANN 9.0.0, driver 25.5.2, TP=4, device 0-3.
